Update module 10

nbarlowATI opened the pull request:
* Minor typo fixes and very small text changes in 10_00, 10_01, 10_02.
* Move 10_03 (domain specific languages) to optional 10_x0
* Move 10_04 (markup languages) to 10_03
* Move 10_05 (controlled vocabularies) to optional 10_x1
* Move 10_06 (semantic models) to optional 10_x2
* Add new notebook 10_04 (beyond pandas, covering row-wise vs column-wise, compression, and the formats arrow, parquet, feather), and corresponding csv datafile match_results.csv

nbarlowATI commented:
There are a couple of open questions with notebooks 5 and 6 in my opinion.  The big ones are:
* "Parallelisation" is not really "scientific file formats" (though it is something that has been requested by previous years' students)
* "Geospatial data" is a bit of a specialist domain - is it appropriate to have this in a generalist RSE course?  

And then there are some details:
* I'm not super confident in the map/reduce example - I looked at a few examples of MapReduce wordcount on the web, and none of them seemed great, this is my cobbled-together understanding, but the "shuffle" stage is carrying a lot of weight!
* There are a couple of things that could potentially be made into Exercises:
   - The MapReduce example in 10.5 - could ask the students to write the functions, given the expected inputs and outputs.
   - The netCDF example in 10.6 - could ask the students to investigate how to retrieve and plot the data from the example file.

mhauru commented:
Comments as I read through the new notebooks: The dask example is clear and good, but I struggled to understand the MapReduce one. Questions I wasn't clear on:
* What are the points where parallelisation would happen if this was run on a cluster, and what parts would be run locally on each node? Some bits are clear, like the loop over files, but this one surprised me: `shuffle_outputs = shuffler(shuffle_outputs)`
* Why is the map part doing the gathering of the counts in such an awkward way using these singleton `dict`s? The answer probably follows from the answer to the above question.
* Why the `sorted(words)`? It seems like the computational bottleneck.
* In what ways does this resemble and not resemble actually using Hadoop? In a real use case, would I write the `mapper`, `shuffler`, and `reducer` functions as done here, and then call some Hadoop function with those as arguments?

mhauru commented:
The geospatial one is concise and easy to follow. A couple of small points:
* Might put in a legend for the plot, to show what the colour scale means.
* I found the last sentence hard to sparse: "Jupyter: interactive notebooks such as Pangeo are a convenient way for users to interact with computing resources, whether running on their local machine, or on a "hub" hosted on the cloud or an HPC cluster, the user experience, and the code, can be almost exactly the same."

nbarlowATI commented:
Fully agree with all those points Markus.   Based on these, and on subsequent in-office discussion, a couple more changes (pending more expert input on MapReduce):
* Added a diagram on MapReduce, which does seem to show the "global shuffle", even though this is not really clear in the other code examples I've seen...
* Added a colour bar to the plot in 10.6.
* Improved the last sentence (about Jupyter) in 10.6.

jemrobinson commented:
- added the new dependencies to `pyproject.toml`
- regenerated `requirements.txt`
- removed `pip install` commands

jemrobinson reviewed (approved):
Thank you, I think this looks very good @nbarlowATI !

jemrobinson commented:
@nbarlowATI @mhauru I'm happy to merge this, are you?

mhauru commented:
Am happy, thanks Nick!
